File: Advanced Techniques/Dynamic_Connectivity.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;

#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vector<int>>;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")

template <typename T>
using ordered_multiset = tree<
    std::pair<T, int>,
    null_type,
    std::less<std::pair<T, int>>,
    rb_tree_tag,
    tree_order_statistics_node_update
>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<typename T>
class is_printable {
    template<typename U> static auto test(int) -> decltype(cout << declval<U>(), true_type{});
    template<typename> static auto test(...) -> false_type;
public: static constexpr bool value = decltype(test<T>(0))::value;
};
template<typename T> enable_if_t<is_printable<T>::value> __print(const T& val) { cout << val; }
template<typename T1, typename T2> void __print(const pair<T1, T2>& p) {
    cout << "("; __print(p.first); cout << ":"; __print(p.second); cout << ")";
}
template<typename T> enable_if_t<!is_printable<T>::value> __print(const T& container) {
    cout << "{"; bool first = true;
    for (const auto& x : container) { if (!first) cout << ", "; __print(x); first = false; }
    cout << "}";
}
void print() { cout << "\n"; }
template<typename T, typename... Args>
void print(const T& t, const Args&... rest) {
    __print(t); if constexpr (sizeof...(rest)) cout << " "; print(rest...);
}

int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable, SortedList: sortl
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree, Trie/Treap: Tries
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, Centroid Decomp.: graph_decom
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, GrapthOth: graphoth, DSU: DSU, FFT:fft
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming


struct RollbackDSU {
    vector<int> parent, sz;
    vector<tuple<int,int,int>> st;
    int components;

    RollbackDSU(int n) {
        parent.resize(n);
        sz.assign(n, 1);
        iota(parent.begin(), parent.end(), 0);
        components = n;
    }

    int find(int x) {
        while (x != parent[x])
            x = parent[x];
        return x;
    }

    void unite(int a, int b) {
        a = find(a);
        b = find(b);

        if (a == b) {
            st.emplace_back(-1, -1, -1);
            return;
        }

        components--;

        if (sz[a] > sz[b])
            swap(a, b);

        st.emplace_back(a, b, sz[b]);
        parent[a] = b;
        sz[b] += sz[a];
    }

    void rollback() {
        auto [a, b, s] = st.back();
        st.pop_back();

        if (a == -1) return;

        components++;
        parent[a] = a;
        sz[b] = s;
    }
};

struct ParPersistentDSU {
    vector<int> parent, sz;
    vector<int> time;

    ParPersistentDSU(int n) {
        parent.resize(n);
        sz.assign(n, 1);
        time.assign(n, INT_MAX);
        iota(parent.begin(), parent.end(), 0);
    }

    int find(int node, int version) {
        while (!(parent[node] == node || time[node] > version)) {
            node = parent[node];
        }
        return node;
    }

    bool unite(int a, int b, int t) {
        a = find(a, t);
        b = find(b, t);

        if (a == b)
            return false;

        if (sz[a] > sz[b])
            swap(a, b);

        parent[a] = b;
        time[a] = t;
        sz[b] += sz[a];

        return true;
    }

    bool isConnected(int a, int b, int version) {
        return find(a, version) == find(b, version);
    }
};

void solve() {
    int n,m,k;
    cin>>n>>m>>k;
    int u,v;

    map<pair<int,int>,int> d;

    while(m--){
        cin>>u>>v;
        u--;v--;
        if(u<v){
            swap(u,v);
        }
        d[{u,v}] = 0;
    }

    vector<tuple<int,int,int,int>> ranges;

    int typ;
    range(i,k){
        cin>>typ>>u>>v;
        u--;
        v--;
        if(u<v){
            swap(u,v);
        }
        if(typ==1){
            d[{u,v}] = i+1;
        }
        else{
            ranges.emplace_back(d[{u,v}],i+1,u,v);
            d.erase(d.find({u,v}));
        }
    }

    for(auto [x,y]:d){
        ranges.emplace_back(y,k+1,x.first,x.second);
    }

    static int base = k+1;
    vector<vector<pair<int,int>>> seg(2*base+1);

    for(auto [l,r,u,v]:ranges){
        int left = l+base;
        int right = r+base;
        while(left<right){
            if(left&1){
                seg[left].emplace_back(u,v);
                left++;
            }
            if(right&1){
                right--;
                seg[right].emplace_back(u,v);
            }
            left >>= 1;
            right >>= 1;
        }
    }

    auto rec = [](int cur,auto &rec,vi &ans,vector<vector<pi>> &tree,RollbackDSU &rds){
        if(cur>=len(tree)){
            return;
        }
        for(auto [u,v]:tree[cur]){
            rds.unite(u,v);
            // print(cur,u,v);
        }
        rec(2*cur,rec,ans,tree,rds);
        rec(2*cur+1,rec,ans,tree,rds);
        if(cur-base>=0 && cur-base<ans.size()){
            ans[cur-base] = rds.components;
        }
        for(auto [u,v]:tree[cur]){
            rds.rollback();
        }
    };

    RollbackDSU rds(n);

    vi ans(k+1);
    rec(1,rec,ans,seg,rds);

    for(auto it:ans){
        cout<<it<<" ";
    }

}

int32_t main() {
    fastio
    int t=1;
    // cin >> t;
    while (t--) solve();
    return 0;
}

File: Advanced Techniques/Dynamic_Connectivity.py
import sys,math,cmath,random,os
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations,combinations
from io import BytesIO, IOBase
from decimal import Decimal,getcontext

BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)

# functions #
# MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(1,2**62)
def gcd(a,b):
    while b:
        a,b = b,a%b
    return a
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
II = lambda : int(sys.stdin.readline().strip())
LII = lambda : list(map(int, sys.stdin.readline().split()))
MI = lambda x : x(map(int, sys.stdin.readline().split()))
SI = lambda : sys.stdin.readline().strip()
SLI = lambda : list(map(lambda x:ord(x)-97,sys.stdin.readline().strip()))
LII_1 = lambda : list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
LII_C = lambda x : list(map(x, sys.stdin.readline().split()))
MATI = lambda x : [list(map(int, sys.stdin.readline().split())) for _ in range(x)]
##

#String hashing: shclass, fenwick sortedlist: fsortl, Number: numtheory/numrare, SparseTable: SparseTable
#Bucket Sorted list: bsortl, Segment Tree(lp,selfop): SegmentTree, bootstrap: bootstrap, Trie: tries
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, BitArray: bitarray
#Combinatorics: pnc, Diophantine Equations: dpheq, DSU: DSU, Geometry: Geometry, FFT: fft, XOR_dict: xdict
#Persistent Segment Tree: perseg, Binary Trie: b_trie, HLD: hld, String funcs: sf, Segment Tree(lp): SegmentOther
#Graph1(dnc,bl): graphadv, Graph2(khn,sat): 2sat, Graph3(fltn,bprt): graphflatten, Graph4(ep,tp,fw,bmf): graphoth
#Graph5(djik,bfs,dfs): graph, Graph6(dfsin): dfsin, utils: utils, Persistent DSU: perdsu, Merge Sort Tree: sorttree
#2-D BIT: 2DBIT, MonoDeque: mono, matrix: nummat, SuffixAutomaton: sautomaton, linalg: linalg, SquareRtDecomp: sqrt
#Grapth7(bridges): graph_dmgt, FWHT(^,|,&): fwht, Graph8(centr_decom): graph_decom, DpOptimize(knth,dnc): dpopt
#2D-SegmentTree: seg2d, Rollback/Par DSU: rbdsu
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
#if os.environ.get('LOCAL'):sys.stdin = open(r'input.txt', 'r');sys.stdout = sys.stderr = open(r'output.txt','w')
#if os.environ.get('LOCAL'):import hashlib;print('Hash Value :',hashlib.md5(open(__file__, 'rb').read()).hexdigest());

class RollbackDSU:
    def __init__(self,n):
        self.p=list(range(n)); self.sz=[1]*n; self.st=[]
        self.size = n
    def find(self,x):
        while x!=self.p[x]: x=self.p[x]
        return x
    def union(self,a,b):
        a,b=self.find(a),self.find(b)
        if a==b: self.st.append((-1,-1,-1)); return
        self.size -= 1
        if self.sz[a]>self.sz[b]: a,b=b,a
        self.st.append((a,b,self.sz[b]))
        self.p[a]=b; self.sz[b]+=self.sz[a]
    def rollback(self):
        a,b,s=self.st.pop()
        if a==-1: return
        self.size += 1
        self.p[a]=a; self.sz[b]=s

class ParPersistentDSU:
    # Partially Persistent DSU with no branching
    def __init__(self,n):
        self.parent = list(range(n))
        self.size = [1]*n
        self.time = [float('inf')]*n
    def find(self,node,version):
        # returns root at given version
        while not (self.parent[node]==node or self.time[node]>version):
            node = self.parent[node]
        return node
    def union(self,a,b,time):
        # merges a and b
        a = self.find(a,time)
        b = self.find(b,time)
        if a==b:
            return False
        if self.size[a]>self.size[b]:
            a,b = b,a
        self.parent[a] = b
        self.time[a] = time
        self.size[b] += self.size[a]
        return True
    def isconnected(self,a,b,time):
        return self.find(a,time)==self.find(b,time)

class OfflineDynamicConnectivity:
    def __init__(self,n):
        self.n = n
        self.active = {}
        self.ranges = []
        self.max_time = 0
    def add_edge(self,u,v,t):
        if u>v:
            u,v = v,u
        self.active[(u,v)] = t
        self.max_time = max(self.max_time,t)
    def remove_edge(self,u,v,t):
        if u>v:
            u,v = v,u
        start = self.active.pop((u,v))
        self.ranges.append((start,t,u,v))
        self.max_time = max(self.max_time,t)
    def run(self):
        dsu = RollbackDSU(self.n)
        for (u,v), start in self.active.items():
            self.ranges.append((start,self.max_time+1,u,v))
        T = self.max_time + 1
        base = T
        tree = [[] for _ in range(2 * base)]
        for l,r,u,v in self.ranges:
            l += base
            r += base
            while l<r:
                if l&1:
                    tree[l].append((u,v))
                    l += 1
                if r&1:
                    r -= 1
                    tree[r].append((u,v))
                l >>= 1
                r >>= 1
        ans = [None] * T
        stack = [(1,0)]
        while stack:
            node,state = stack.pop()
            if state==0:
                for u,v in tree[node]:
                    dsu.union(u,v)
                stack.append((node,1))
                if node<base:
                    stack.append((node*2+1,0))
                    stack.append((node*2,0))
                else:
                    t = node-base
                    if t<T:
                        ans[t] = dsu.size
                        # implement a size function
                        # size-- on union
            else:
                for _ in tree[node]:
                    dsu.rollback()
        return ans
    
def solve():
    n,m,k = LII()
    
    odc = OfflineDynamicConnectivity(n)

    for _ in range(m):
        u,v = LII_1()
        odc.add_edge(u,v,0)

    for _ in range(k):
        t,u,v = LII_1()
        if t==0:
            odc.add_edge(u,v,_+1)
        else:
            odc.remove_edge(u,v,_+1)
    # print(ranges)
    ans = odc.run()
    print(*ans)

    #L1 = LII()
    #st = SI()
solve()